gnomad_constraint/utils/constraint_basics.py

```diff
@@ -165,20 +167,221 @@ def get_proportion_observed_by_coverage(
+            - mu_snp - mutation rate
+            - possible_variants - the number of possible variants in the dataset for each substitution (`alt`), `context`,
+            `methylation_level`, and median exome coverage (`exome_coverage`)
+
+    :param coverage_ht: Input coverage Table.
+    :param trimers: Whether the contexts were trimmed or not. Defaults to False.
+    :param weighted: Whether to weight the high coverage model (a linear regression model) by 'possible_variants'. Defaults to False.
+    :param pop_anal: Whether to build models for each population. Defaults to True.
+    :param pops: List of populations. Defaults to `POPS`.
+    :param cov_cutoff: A coverage cutoff. Coverage of sites above the cutoff will be used to build plateau models. Otherwise they will be used to build coverage models. defaults to `HIGH_COVERAGE_CUTOFF`.
+    :return: Coverage model and plateau models.
     """
-    annotation = t.vep.annotate(
-        transcript_consequences=t.vep.transcript_consequences.map(
-            add_most_severe_consequence_to_consequence
+    keys = ["context", "ref", "alt", "methylation_level", "mu_snp"]
+
+    all_high_coverage_ht = coverage_ht.filter(coverage_ht.exome_coverage >= cov_cutoff)
+    agg_expr = {
+        "observed_variants": hl.agg.sum(all_high_coverage_ht.variant_count),
+        "possible_variants": hl.agg.sum(all_high_coverage_ht.possible_variants),
+    }
+    if pop_anal:
+        for pop in pops:
+            agg_expr[f"observed_{pop}"] = hl.agg.array_sum(
+                all_high_coverage_ht[f"downsampling_counts_{pop}"]
+            )
+    high_coverage_ht = all_high_coverage_ht.group_by(*keys).aggregate(**agg_expr)
+
+    high_coverage_ht = annotate_variant_types(high_coverage_ht, not trimers)
+    plateau_models = build_plateau_models(high_coverage_ht, weighted=weighted)
+
+    high_coverage_scale_factor = all_high_coverage_ht.aggregate(
+        hl.agg.sum(all_high_coverage_ht.variant_count)
+        / hl.agg.sum(
+            all_high_coverage_ht.possible_variants * all_high_coverage_ht.mu_snp
         )
     )
-    return (
-        t.annotate_rows(vep=annotation)
-        if isinstance(t, hl.MatrixTable)
-        else t.annotate(vep=annotation)
+
+    all_low_coverage_ht = coverage_ht.filter(
+        (coverage_ht.exome_coverage < cov_cutoff) & (coverage_ht.exome_coverage > 0)
+    )
+
+    low_coverage_ht = all_low_coverage_ht.group_by(
+        log_coverage=hl.log10(all_low_coverage_ht.exome_coverage)
+    ).aggregate(
+        low_coverage_obs_exp=hl.agg.sum(all_low_coverage_ht.variant_count)
+        / (
+            high_coverage_scale_factor
+            * hl.agg.sum(
+                all_low_coverage_ht.possible_variants * all_low_coverage_ht.mu_snp
+            )
+        )
+    )
+    coverage_model = build_coverage_model(low_coverage_ht)
+    # TODO: consider weighting here as well
+
+    return coverage_model, plateau_models
+
+
+def build_plateau_models(
```

Comment:
Do you mean merge this with `build_plateau_models_pop` and `build_plateau_models_total`? I separate them because the structure will be much more clear and readable.

Reply:
I feel the code is so similar that these functions could easily be combined into one function though